### PR TITLE
Issue66

### LINKED ISSUE #66: Update PUT /donatedItem/status/ Endpoint to Handle Image Uploads
Enhance the **PUT** `/donatedItem/status/:id` endpoint to handle the uploading of image files, similar to the **POST** `/donatedItem` endpoint. The updated endpoint should accept image files from the client, upload these to cloud storage, and store the resultant URLs in the database alongside other status details.

**Tasks:**

- [ ] Image File Handling:
  - Modify the endpoint to receive imageFiles as a multipart file format.
  - Utilize Express.Multer.File array to process the incoming image files.
   
- [ ] Image Upload to Cloud Storage:
  - Implement functionality to upload each image file to cloud storage.
  - Generate a unique filename for each image based on the donated item ID and the current timestamp to avoid naming conflicts.
  - Retrieve URLs of the uploaded images.

- [ ] Database Update:
  - Update the corresponding record in the donatedItemStatus table with the new statusType, dateModified, and imageUrls.
  - Handle potential errors and conflicts when updating the record.

- [ ] Response Handling:
  - Upon successful update, return a 200 HTTP status with a JSON object including updated status details and image URLs.
  - In case of errors (validation, storage, database update), return appropriate HTTP status codes with error messages.

**Developer Notes:**
Refer to the implementation in the **POST** `/donatedItem` route as a template for handling file uploads and interactions with cloud storage. Specifically, examine how files are named uniquely and uploaded, and how URLs are subsequently stored in the database. The code at https://github.com/oss-slu/material-donor-mutual-assist/blob/main/server/src/routes/donatedItemRoutes.ts provides a practical reference.

### PR DESCRIPTION
**Pull request by miabu-pashh**
Fixes #66 

### What was changed?

- Enhanced the PUT /donatedItem/status/:id endpoint to handle the uploading of image files.

- Modified the endpoint to:

- Accept imageFiles  in a multipart file format.

- Upload image files to cloud storage and generate unique URLs.

- Update the donatedItemStatus table with the statusType, dateModified, and imageUrls.

- Added error handling for validation, file uploads, and database updates.

### Why was it changed?

- Previously, the PUT /donatedItem/status/:id endpoint lacked the capability to handle image file uploads.

- The enhancement was needed to allow users to upload images associated with a status update and store their URLs for future reference.

- This improvement ensures a better user experience and provides more flexibility for managing donated items and their statuses.

### How was it changed?

- Updated the endpoint logic in the donatedItemRoutes.ts file:

- Added upload.array('imageFiles') middleware to process multipart file uploads.

- Integrated the uploadToStorage function to upload image files to cloud storage and generate unique filenames.

- Used prisma.donatedItemStatus.update() to update the status record in the database with the new details and image URLs.

- Refactored validation to ensure statusType and dateModified are mandatory fields.

### Screenshots that show the changes (if applicable):
Postman Success Message:
Before:

The endpoint did not support image uploads and returned an error if images were included.
After:
Successfully updated the donatedItemStatus with image URLs:

Screenshot 2024-11-11 at 12 22 01 AM

### CODE REVIEW

File: server/src/routes/donatedItemRoutes.ts [hunk 1]
```diff
@@ -145,43 +145,74 @@ router.get('/:id', async (req: Request, res: Response) => {
...
+
+            // Extract status details and image files
+            const { statusType, dateModified } = req.body;
+            const imageFiles = req.files as Express.Multer.File[];
+
+            // Ensure statusType and dateModified are provided
+            if (!statusType || !dateModified) {
+                return res
+                    .status(400)
+                    .json({ error: 'statusType and dateModified are required' });
```
Comment (saikiran0405) [inline server/src/routes/donatedItemRoutes.ts:167]: Enhance validation for statusType and dateModified to check for valid formats or values.

File: server/src/routes/donatedItemRoutes.ts [hunk 2]
```diff
@@ -145,43 +145,74 @@ router.get('/:id', async (req: Request, res: Response) => {
...
             }
 
-            const updatedItem = await prisma.donatedItem.update({
-                where: { id: Number(req.params.id) },
+            // Convert dateModified to a Date object
+            const dateModifiedDateTime = new Date(dateModified);
+            dateModifiedDateTime.setUTCHours(0, 0, 0, 0); // Normalize time to UTC
+
+            // Upload images to cloud storage and retrieve URLs
```
Comment (saikiran0405) [inline server/src/routes/donatedItemRoutes.ts:174]: Ensure there's a check for imageFiles being undefined or empty before processing uploads.

File: server/src/routes/donatedItemRoutes.ts [hunk 3]
```diff
@@ -145,43 +145,74 @@ router.get('/:id', async (req: Request, res: Response) => {
...
+                status: updatedStatus,
             });
+        } catch (error) {
+            console.error('Error updating donated item status:', error);
+
+            if (error instanceof Error) {
+                res.status(400).json({ error: error.message });
+            } else {
+                res.status(500).json({
+                    message: 'Error updating donated item status',
```
Comment (saikiran0405) [inline server/src/routes/donatedItemRoutes.ts:208]: Consider refining error handling to address specific exceptions more comprehensively

**Review by saikiran0405 — COMMENTED** [review]
1. In the code review for the `PUT /donatedItem/status/:id` endpoint, I have addressed enhancements in error handling, input validation, and date processing. 

2. Recommendations included refining error management, improving the validation of input parameters for accuracy, and revising the approach to date normalization. 

3. Additionally, suggested implementing safeguards against potential file handling errors

File: server/src/routes/donatedItemStatusRoutes.ts [hunk 4]
```diff
@@ -2,15 +2,46 @@ import { Router, Request, Response } from 'express';
...
         try {
+            const donatedItemId = Number(req.params.id);
+
             const { statusType, dateModified } = req.body;
+            const imageFiles = req.files as Express.Multer.File[];
+
+            if (!statusType) {
+                return res.status(400).json({ error: 'status is required' });
+            }
+            const imageUrls = await Promise.all(
```
Comment (Anjali0407-git) [inline server/src/routes/donatedItemStatusRoutes.ts:30]: Does it works even when imageFiles is sent as null?

File: server/src/routes/donatedItemStatusRoutes.ts [hunk 5]
```diff
@@ -2,15 +2,46 @@ import { Router, Request, Response } from 'express';
...
+                    return await uploadToStorage(file, uniqueFilename);
+                }),
+            );
+
+            const updatedItem = await prisma.donatedItem.update({
+                where: { id: donatedItemId },
+                data: {
+                    currentStatus: statusType,
+                    lastUpdated: new Date(),
+                },
```
Comment (Anjali0407-git) [inline server/src/routes/donatedItemStatusRoutes.ts:43]: Already prisma create and update methods are in place [as shown ](https://github.com/oss-slu/material-donor-mutual-assist/blob/main/server/src/routes/donatedItemStatusRoutes.ts#L16C12-L33C16). Just adding `imageUrls `to .update method would work.

**Review by Anjali0407-git — CHANGES_REQUESTED** [review]

**Comment by miabu-pashh** [comment]
image

**Review by Anjali0407-git — APPROVED** [review]

**Pull request merged**